File: src/sunstone/views/tabs/audit_tab.rb
# Sunstone tab for audit trail
# Register in sunstone views/tabs tab section

module OpenNebula
  class SunstoneViews
    def self.audit_tab
      {
        name: 'Audit Trail',
        view: 'audit',
        icon: 'fa-history',
        html: <<-HTML
<div id="audit_tab" class="row">
  <div class="col-md-12">
    <form id="audit_filter" class="form-inline">
      <div class="form-group">
        <label for="user_id">User ID:</label>
        <input type="text" id="user_id" class="form-control" placeholder="User ID">
      </div>
      <div class="form-group">
        <label for="start_time">Start:</label>
        <input type="datetime-local" id="start_time" class="form-control">
      </div>
      <div class="form-group">
        <label for="end_time">End:</label>
        <input type="datetime-local" id="end_time" class="form-control">
      </div>
      <div class="form-group">
        <label for="request_name">Request:</label>
        <input type="text" id="request_name" class="form-control" placeholder="e.g., one.vm.deploy">
      </div>
      <div class="form-group">
        <label for="object_id">Object ID:</label>
        <input type="text" id="object_id" class="form-control" placeholder="Object ID">
      </div>
      <button type="button" id="audit_search" class="btn btn-primary">Search</button>
    </form>
    <hr>
    <table id="audit_results" class="table table-striped">
      <thead>
        <tr>
          <th>Timestamp</th>
          <th>User ID</th>
          <th>Request</th>
          <th>Params</th>
          <th>Result</th>
          <th>Objects</th>
        </tr>
      </thead>
      <tbody>
      </tbody>
    </table>
  </div>
</div>
<script>
$('#audit_search').click(function() {
  var filters = {
    user_id: $('#user_id').val(),
    start: $('#start_time').val(),
    end: $('#end_time').val(),
    request: $('#request_name').val(),
    object_id: $('#object_id').val()
  };
  $.ajax({
    url: '/audit/query',
    type: 'POST',
    data: JSON.stringify(filters),
    contentType: 'application/json',
    success: function(data) {
      var tbody = $('#audit_results tbody');
      tbody.empty();
      data.forEach(function(entry) {
        tbody.append('<tr><td>' + entry.timestamp + '</td><td>' + entry.user_id + '</td><td>' + entry.request + '</td><td>' + entry.params + '</td><td>' + entry.result + '</td><td>' + entry.objects + '</td></tr>');
      });
    }
  });
});
</script>
        HTML
      }
    end
  end
end

File: sunstone/js/audit_tab.js
// Sunstone audit trail tab
(function($) {
  $.fn.auditTab = function() {
    var container = this;
    container.html('<h2>Audit Trail</h2>' +
      '<div id="audit-filters">' +
        'User ID: <input type="number" id="filter-user" /> ' +
        'Action: <input type="text" id="filter-action" /> ' +
        '<button id="filter-btn">Filter</button>' +
      '</div>' +
      '<table id="audit-table" border="1">' +
        '<thead><tr><th>Timestamp</th><th>User</th><th>Action</th><th>Arguments</th><th>Result</th><th>Objects</th></tr></thead>' +
        '<tbody></tbody>' +
      '</table>');

    function loadData(params) {
      $.getJSON('/audit/list', params || {}, function(data) {
        var tbody = container.find('#audit-table tbody');
        tbody.empty();
        data.forEach(function(row) {
          tbody.append('<tr>' +
            '<td>' + row.timestamp + '</td>' +
            '<td>' + row.user_id + '</td>' +
            '<td>' + row.action + '</td>' +
            '<td>' + (row.arguments ? JSON.parse(row.arguments) : '') + '</td>' +
            '<td>' + (row.result ? JSON.parse(row.result) : '') + '</td>' +
            '<td>' + (row.objects ? JSON.parse(row.objects) : '') + '</td>' +
            '</tr>');
        });
      });
    }

    container.find('#filter-btn').click(function() {
      var params = {};
      var userId = container.find('#filter-user').val();
      if (userId) params.user_id = userId;
      var action = container.find('#filter-action').val();
      if (action) params.action = action;
      loadData(params);
    });

    loadData();
  };
})(jQuery);

// Initialize when document ready
$(document).ready(function() {
  $('#audit-tab-container').auditTab();
});